
```{r setup, include=FALSE, eval = TRUE, echo = FALSE}
opts_chunk$set(
  echo=input$report_echo,
  error=TRUE,
  dpi = 100
)
```

```{r evals, echo = FALSE}
evalJoined <- FALSE
evalSampleplan <- FALSE
evalCounts <- FALSE
evalGenesList <- FALSE
if (!is.null(reactives$joined)){evalJoined <- TRUE}
if (!is.null(reactives$sampleplan)){evalSampleplan <- TRUE}
if (!is.null(reactives$countsRaw)){evalCounts <- TRUE}
if(evalJoined == TRUE){
if (!is.null(input$essential) && !is.null(input$nonessential)){
  evalGenesList <- TRUE
}
}
options(kableExtra.auto_format = F)
```


# Data overview

## Files 
```{r, echo = FALSE, eval = evalCounts}
br()
write.table(reactives$countsRaw,file = "/tmp/Rawcounts.csv",sep = ",",quote = FALSE)
br()
xfun::embed_files(path = "/tmp/Rawcounts.csv", text = "Counts Matrix File")
```

```{r, echo = FALSE, eval = evalSampleplan}
br()
write.table(reactives$sampleplan,file = "/tmp/sampleplan.csv",sep = ",",quote = FALSE)
br()
xfun::embed_files(path = "/tmp/sampleplan.csv", text = "SamplePlan file")
```

```{r, echo = FALSE, eval = !evalCounts}
print("Please upload counts matrix first")
```

```{r, echo = FALSE, eval = !evalSampleplan}
print("Please upload both SamplePlan first")
```

## Samples Selected

The following samples were selected in this analysis : 

<button title="Click to show answer" type="button" onclick="if(document.getElementById('BE_Genus_C.txt') .style.display=='none') {document.getElementById('BE_Genus_C.txt') .style.display=''}else{document.getElementById('BE_Genus_C.txt') .style.display='none'}">Show/hide</button>
<div id="BE_Genus_C.txt" style="display:none">
```{r, echo = FALSE, eval = evalSampleplan}
kableExtra::kable(reactives$sampleplan,caption = "samplePlan Table") %>%
kable_styling() %>%
kableExtra::column_spec(1:ncol(reactives$sampleplan),width = "80em", bold = FALSE, italic = FALSE)# %>%
  #scroll_box(width = "100%", height = "20%")

```

```{r, echo = FALSE, eval = !evalSampleplan}
print("Please upload SamplePlan file first")
```
</div>
  
## Distributions

```{r, echo = FALSE, eval = !evalJoined}
print("Please upload both SamplePlan and counts matrix first")
```

### Read Numbers
```{r, echo = FALSE, eval = evalJoined}

  counts <- reactives$joined
      counts <- counts %>%
        group_by(.data$Sample_ID, .data$Replicate, .data$Cell_line, .data$Timepoint) %>%
        summarise(total = sum(.data$count)) %>% 
        as.data.frame() %>%
        ggplot(aes(x = .data$Timepoint, y = .data$total)) +
        geom_col(position = position_dodge()) + facet_wrap(vars(.data$Replicate, .data$Cell_line), nrow = 1) +
        labs(title = "Number of reads by sample") #+

plot(counts)    
```
<br/><br/>

### Log cpm distributions
<br/>

```{r, echo = FALSE, eval = evalJoined}
    counts <- reactives$joined
      counts %>% 
        ggplot(aes(x = .data$Timepoint, y = .data$log_cpm, fill = .data$Replicate)) + geom_boxplot() + facet_grid(. ~ .data$Cell_line) + 
        labs(title = "Distribution of normalized log-cpm by sample", subtitle = "All guides")
```

<br/><br/>

### Density ridges

#### All guides
<br/>
```{r, echo = FALSE, eval = evalJoined, message=FALSE}
 counts <- reactives$joined
      counts %>%
        ggplot(aes(x = .data$log_cpm, y = .data$Timepoint)) +
        geom_density_ridges(alpha = 0.6, show.legend = FALSE, fill = "gray50") +
        facet_wrap(vars(.data$Cell_line, .data$Replicate), ncol = 1, strip.position = "right") +
        labs(title = "Distribution of normalzed log-cpm by sample", subtitle = "All guides")
```
<br/><br/><br/>
#### Per gene type

```{r, echo = FALSE, eval = evalGenesList,message=FALSE}
### Essential genes
counts <- reactives$joined
ess_genes <- ess_genes()
counts <- filter(counts, .data$gene %in% ess_genes$V1)
plot1 <- ggplot(counts, aes(x = .data$log_cpm, y = .data$Timepoint)) + 
  geom_density_ridges(alpha = 0.6, show.legend = FALSE, fill = "gray50") +
  facet_wrap(vars(.data$Cell_line, .data$Replicate), ncol = 1, strip.position = "right") +
  scale_fill_viridis_c() +
  labs(title = "Distribution of normalised log-cpms", subtitle = "Essential genes")

# Non essential genes
counts <- reactives$joined
ess_genes <- non_ess_genes()
counts <- counts %>%
  filter(.data$gene %in% ess_genes$V1)
plot2 <- ggplot(counts, aes(x = .data$log_cpm, y = .data$Timepoint)) +
  geom_density_ridges(alpha = 0.6, show.legend = FALSE, fill = "gray50") +
  facet_wrap(vars(.data$Cell_line, .data$Replicate), ncol = 1, strip.position = "right") +
  scale_fill_viridis_c() +
  labs(title = "Distribution of normalised log-cpms", subtitle = "Non Essential genes")
```

```{r, echo = FALSE, eval = evalGenesList,message=FALSE,figures-side, fig.show="hold"}
grid.arrange(plot1,plot2,ncol = 2)
```
<br/>

```{r, echo = FALSE, eval = !evalGenesList}
print("Please provide essential and non essential genes lists first")
```

## Results

### Difference to initial timepoint

```{r, echo = FALSE, eval = !evalGenesList}
print("Please provide essential and non essential genes lists first")
```

##### All guides

```{r, echo = FALSE, eval = evalGenesList}

if(is.null(diff_boxes$diff_box_all) | is.null(diff_boxes$diff_box_ess)){
  print("Please perform positive screening in the dedicated app tab first")
} else { 
plot(diff_boxes$diff_box_all)
}
```

#### Essential genes

```{r, echo = FALSE, eval = evalGenesList}

if(is.null(diff_boxes$diff_box_all) | is.null(diff_boxes$diff_box_ess)){
  print("Please perform positive screening in the dedicated app tab first")
} else { 
plot(diff_boxes$diff_box_ess)
}
```


# About COOKIE CRISP'R

`COOKIE CRISP'R` is a package containing a Shiny application for
analyzing CRISPR screening experiments data in different conditions and experimental factors.

`COOKIE CRISP'R` was developed by the bio-informatic platform of the Curie's Institute (France). 

## Developers

`COOKIE CRISP'R` is currently maintained by Cl√©ment BENOIT and Pierre Gestraud at the Bioinformatic Platform (Institut Curie) (https://curie.fr/).
You can contact them by clicking on the buttons below.

<a href="mailto:mailto:pierre.gestraud@curie.fr?subject=[COOKIE CRISP'R_feedback]" class="btn btn-primary">Pierre Gestraud</a>

<a href="mailto:mailto:clement.benoit@curie.fr?subject=[COOKIE CRISP'R_feedback]" class="btn btn-primary">Clement BENOIT</a>


# Citation info

If you use `COOKIE CRISP'R` for your analysis, please cite it as here below:

```{r, warning=FALSE, echo = FALSE}
citation("CRISPRApp")
```

# Session Information

```{r,echo=FALSE}
sessionInfo()
```

```{r, echo = FALSE}
library(shiny)
footertemplate <- function(){
  tags$div(
    class = "footer",
    style = "text-align:center",
    tags$div(
      class = "foot-inner",
      list(
        hr(),
        "This report was generated with", tags$a(href="http://bioconductor.org/packages/COOKIE CRISP'R/", "COOKIE CRISP'R"), br(),
        "COOKIE CRISP'R is a project developed by Nicolas Servant, Pierre Gestraud, Aurelien Bore and Clement Benoit from the bioinformatic and crispr'it platforms of the Institut Curie : ",
        br(),
        tags$a(href="https://science.curie.fr/plateformes/criblage-genetique-crispr-crisprit/","CRISPRIT platform"),
        br(),
        tags$a(href="https://science.curie.fr/recherche/developpement-cancer-genetique-epigenetique/dynamique-de-linformation-genetique/equipe-vallot/","Bioinformatic Platform"),br())
      )
    )
}
```

```{r, echo = FALSE}
footertemplate()
```
